<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Alejandro Granados" />


<title>An introduction to the analysis single-cell RNA sequencing data</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="index.html">CdC2019</a>
</li>
<li>
  <a href="index.html">Projects</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Workshop
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Intro.html">Day 2: Intro R</a>
    </li>
    <li>
      <a href="scRNAseq.html">Day 3: scRNA seq</a>
    </li>
    <li class="divider"></li>
    <li>
      <a href="metadata.html">Additional Material</a>
    </li>
  </ul>
</li>
<li>
  <a href="https://github.com/rstudio/distill">
    <span class="fab fa fab fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">An introduction to the analysis single-cell RNA sequencing data</h1>
<h4 class="author">Alejandro Granados</h4>
<h4 class="date">24/07/2019</h4>

</div>


<div id="single-cell-rna-seq" class="section level2">
<h2>Single cell RNA seq</h2>
<p>Our bodies and all of the living organisms are made of cells, the fundamental units of life. There are also viruses but let’s keep them for another day. Each cell has a copy of the genetic information necessary to grow, thrive in diverse environments and reproduce. In our bodies, there are around 37 trillion cells. Each one of them has <strong>the same</strong> genetic information. So, why are they different? The cells in our eyes are different from the cells in our skin. The difference relies on their <em>personality</em>. The personality of a cell is determined in part from <em>how</em> they use their genetic information. Which genes are ON or OFF and WHEN?</p>
<p>Inside a cell, there are thousands of reactions taking place in parallel, some of them deal with information. Turn a gene ON means, in general: let’s read the DNA, transcribe the message to RNA and then <em>translated</em> to produce proteins, the building blocks of cells. Therefore, DNA is <em>potential</em> information but needs to be <em>transcribed</em> to RNA to be used or <em>expressed</em> into proteins. In general, RNA molecules tell us which genes are actively being transcribed or <em>expressed</em> and so, having information about RNA levels inside the cell tell us something about the personality of a cell! As with people, personality is determined by multiple factors, for cells gene expression is one important factor but not the only one.</p>
<p>Single-cell RNA-seq examines the sequence information of RNA inside individual cells as a variable that indicates gene activity. By amplifying the signal from RNA molecules inside the cell and converting the RNA into DNA, we can use sequencing methods to recover information about which RNAs were present in a given cell. Moreover, different technologies allow sequencing of multiple cells in parallel. For example, the 10x Genomics platform enables up to 10,000 cells per sample! Therefore, scRNA-seq provides us with data for individual cells of which genes they had turned ON or OFF.</p>
<p>The number of genes in a given cell (~23,000 for mouse) times the number of cells means that we are dealing with really large matrices when analyzing RNA-seq data. One challenge, however, is that until today we cannot read <em>ALL</em> the RNA from a single cell. During the process of sequencing the sample, most of the RNA is lost! This <code>dropout</code> problem imposes serious challenges for data analysis.</p>
<p>Here, we will go through a standard pipeline to analyze scRNA-seq data from a cell atlas: The Tabula Muris <a href="https://tabula-muris.ds.czbiohub.org/" class="uri">https://tabula-muris.ds.czbiohub.org/</a>. In a cell atlas, researchers try to sequence the whole organism with single-cell resolution. As you can imagine, this is a huge amount of data. For this tutorial, we will focus on a single tissue.</p>
<p>We will use the specialized library <code>Seurat</code> which has been developed by Rahul Sajita and co-workers <a href="https://satijalab.org/seurat/" class="uri">https://satijalab.org/seurat/</a></p>
<p>Let’s start by downloading the dataset from here: <a href="https://figshare.com/articles/Single-cell_RNA-seq_data_from_Smart-seq2_sequencing_of_FACS_sorted_cells_v2_/5829687" class="uri">https://figshare.com/articles/Single-cell_RNA-seq_data_from_Smart-seq2_sequencing_of_FACS_sorted_cells_v2_/5829687</a></p>
<pre><code>## 
## Attaching package: &#39;dplyr&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     filter, lag</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     intersect, setdiff, setequal, union</code></pre>
<pre><code>## here() starts at /Users/alejandrog/DataScienceMX</code></pre>
</div>
<div id="loading-data" class="section level2">
<h2>Loading Data</h2>
<p>We can read the data using the function <code>read.csv</code> and convert it into a sparse matrix (data is full of zeros!)</p>
<p>Sort the cells by name (their name contains information about the experimental conditions)</p>
<p>Now, let’s load the <em>metadata</em>, which contains the information about the data!</p>
<pre><code>##   plate.barcode mouse.id  tissue subtissue FACS.selection mouse.sex
## 1       D041914    3_8_M Bladder                 Multiple         M
## 2       D042253    3_9_M Bladder                 Multiple         M
## 3     MAA000487   3_10_M Bladder                 Multiple         M
## 4       B000610   3_56_F Bladder                 Multiple         F
## 5       B002764   3_38_F Bladder                 Multiple         F
## 6       B002771   3_39_F Bladder                 Multiple         F</code></pre>
<p>The <em>metadata</em> has all the information about the experiments. Every organ was processed and sequenced by a different research group and so we need to find <strong>which cells correspond to which experiment</strong>. In the <a href="https://www.nature.com/articles/s41586-018-0590-4">protocol</a> they use for scRNA-seq, each cell is sequenced on a well of a <a href="https://en.wikipedia.org/wiki/Microplate">plate</a>. The information in the metadata represents plates! Let’s link the <strong>plates</strong> to the <strong>cells</strong></p>
<p>The ERCC sequences tell us about the library size and differences between cell to cell. We can plot the amount of ERCC in the data per cell and see if there are clear outliers that we should remove.</p>
<pre class="r"><code>hist(percent.ercc,main=&quot;Percentage of control ERCC reas per cell&quot;,xlab = &quot;% ERCC per cells&quot;)</code></pre>
<p><img src="scRNAseq_files/figure-html/ercc-1.png" width="672" /></p>
<p>Since a cell in principle should not have too many ERCC (these are only for control) we should remove cells with high ERCC content</p>
<p>ERCC sequences are useful as a <em>control</em> for the data because they were added by the scientists in <em>known</em> concentrations. In this way we expect ALL cells to have the same ERCC content, right? But things are not that simple, the technology is not perfect and different cells are sequenced in different ways or with different efficiency. Can you think of a way in which we could use ERCC sequences to <em>normalize</em> the data?</p>
<p>First, we want to get rid of cell with <em>too</em> high ERCC content. Where would you put a <em>cutoff</em>?</p>
<pre class="r"><code>raw.data = raw.data[,percent.ercc&lt;0.15]</code></pre>
<pre class="r"><code>#Take from meta.data only cells that remain in the matrix

cell.meta.filter = cell.meta.data[colnames(raw.data), ]

head(cell.meta.filter)</code></pre>
<pre><code>##                          plate.barcode mouse.id tissue subtissue
## A1.B001717.3_38_F.1.1          B001717   3_38_F Kidney          
## A10.B002775.3_39_F.1.1         B002775   3_39_F Kidney          
## A10.MAA000752.3_10_M.1.1     MAA000752   3_10_M Kidney          
## A11.MAA000801.3_11_M.1.1     MAA000801   3_11_M Kidney          
## A12.B001717.3_38_F.1.1         B001717   3_38_F Kidney          
## A12.MAA000801.3_11_M.1.1     MAA000801   3_11_M Kidney          
##                          FACS.selection mouse.sex
## A1.B001717.3_38_F.1.1            Viable         F
## A10.B002775.3_39_F.1.1           Viable         F
## A10.MAA000752.3_10_M.1.1         Viable         M
## A11.MAA000801.3_11_M.1.1         Viable         M
## A12.B001717.3_38_F.1.1           Viable         F
## A12.MAA000801.3_11_M.1.1         Viable         M</code></pre>
<p>Let’s now check <em>how different</em> are the cells with each other. Are there cells that have not enough data? The columns in the matrix represent the cells so we can sum over the columns and plot a histogram. In this way we can filter <em>bad quality</em> cells.</p>
<pre class="r"><code># Total number of reads per cell
hist(colSums(raw.data),main = &quot;Total number of reads per cell&quot;,xlab =&quot;N Reads per cell&quot; )</code></pre>
<p><img src="scRNAseq_files/figure-html/ercc3_1-1.png" width="672" /></p>
<pre class="r"><code># But some genes might have many reads while some other genes might have low expression in living cells. 
# We can also ask then, how many genes have at least 1 read 

hist(colSums(raw.data&gt;0),main = &quot;Total number of genes expressed by cell&quot;,xlab = &quot;Genes per cell&quot;)</code></pre>
<p><img src="scRNAseq_files/figure-html/ercc3_1-2.png" width="672" /></p>
<p>What is the average number of genes expressed per cells?</p>
<pre class="r"><code>#Note that we define a gene as expressed if the cell has at least 1 read. But this definition can change. 

mean(colSums(raw.data&gt;0))</code></pre>
<pre><code>## [1] 1665.79</code></pre>
<pre class="r"><code>#Based on the histogram we can keep only cells that have more than 1000 genes with &gt;0 reads
raw.data&lt;-raw.data[,colSums(raw.data&gt;0)&gt;1000]
#Try looking at the histogram now</code></pre>
<p>Anothe imporant filter is by gene. Some genes are very poorly expressed, maybe because they have low expression in general or maybe they are not expressed in this particular tissue/condition.</p>
<p>To see the distribution of genes, we need to sum by row.</p>
<pre class="r"><code>#we can arrange two plots in a single figure using par()
# mrfow=c(1,2) mean we want 1 row and two columns to arrange the plots 
par(mfrow = c(1,2))
# Number of reads for each gene across cells (plot 1)
hist(rowSums(raw.data),main=&quot;Total number of reads per gene &quot;,xlab=&quot;N reads per gene&quot;)
# Let&#39;s look at it on log scale
hist(log10(rowSums(raw.data)),main=&quot;Total number of reads per gene &quot;,xlab=&quot;N reads per gene (log)&quot;)</code></pre>
<p><img src="scRNAseq_files/figure-html/filtergenes-1.png" width="672" /></p>
<pre class="r"><code># We can also look at how many cells express each gene
par(mfrow = c(1,2))
hist(rowSums(raw.data&gt;0),main=&quot;Number of expressing cells per gene&quot;)

hist(log10(rowSums(raw.data&gt;0)),main=&quot;Number of expressing cells per gene (log)&quot;)</code></pre>
<p><img src="scRNAseq_files/figure-html/filtergenes-2.png" width="672" /> As we can see most genes are expressed only by a few cells</p>
<div id="seurat" class="section level3">
<h3>Seurat</h3>
<p>Seurat is a R-based software package for analysis of single cell RNA seq. I has been shown to perfom well across different datasets and normalization methods.</p>
<p>Here are some references where people has compared Seurat with other methods (there are many methods and packages for analysis of single cell RNA seq!) <a href="https://www.nature.com/articles/s41592-019-0425-8" class="uri">https://www.nature.com/articles/s41592-019-0425-8</a></p>
<pre class="r"><code># # # # ## SEURAT OBJECT
tissue &lt;- CreateSeuratObject(counts = raw.data)</code></pre>
<pre><code>## Warning: Feature names cannot have underscores (&#39;_&#39;), replacing with dashes
## (&#39;-&#39;)</code></pre>
<pre class="r"><code>tissue &lt;- AddMetaData(object = tissue, cell.meta.data)
tissue &lt;- AddMetaData(object = tissue, percent.ercc, col.name = &quot;percent.ercc&quot;)</code></pre>
<p>We can now normalize the data. One of the technical problems of RNA sequencing is that some RNA molecules are not detected and other are highly produced in cells, so the <em>distribution</em> of expression is <em>wide</em>.</p>
<pre class="r"><code>  tissue&lt;-NormalizeData(object = tissue, verbose=F)
  tissue&lt;-FindVariableFeatures(object=tissue,selection.method =&quot;vst&quot;,nfeatures =2000)
   tissue&lt;-ScaleData(tissue)</code></pre>
<pre><code>## Centering and scaling data matrix</code></pre>
<p>Cells have 20k genes, but most of them are not expressed and some of them do not change in different cell types. Those are not informative genes. Some genes, however, are specific for given cell types. The PCA will find those genes and create <em>principal components</em> which are linear combinations of genes that explain the variation between cells.</p>
<pre class="r"><code>tissue &lt;- RunPCA(object = tissue, npcs = 100, verbose = FALSE)

VizDimLoadings(object = tissue, dims = 1:4, reduction = &quot;pca&quot;)</code></pre>
<p><img src="scRNAseq_files/figure-html/seurat3-1.png" width="672" /></p>
<pre class="r"><code>ElbowPlot(tissue)</code></pre>
<p><img src="scRNAseq_files/figure-html/seurat3-2.png" width="672" /></p>
<p>An elbow plot tells us how many of these <em>dimensions</em> or <em>principal components</em> we need to explain the spread of the data. While the question of how many principal components should we keep is not trivial, the elbow plot gives us an estimate. For example the first 3 principal components explain a lot of standard deviation. After the 7th PC, most other component do not contribute significantly and so we say that we found an “elbow”. There is another elbow at PC 15.</p>
<pre class="r"><code>#Run this only if you have python installed and if you installed UMAP
tissue &lt;- RunUMAP(object = tissue, reduction = &quot;pca&quot;, dims = 1:40)

# A different method for visualization is tSNE
tissue &lt;- RunTSNE(tissue,dims=1:10)

tissue &lt;- FindNeighbors(object = tissue, reduction = &quot;pca&quot;, dims = 1:40)</code></pre>
<pre><code>## Computing nearest neighbor graph</code></pre>
<pre><code>## Computing SNN</code></pre>
<pre class="r"><code>tissue &lt;- FindClusters(tissue, resolution = 0.8)</code></pre>
<pre><code>## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
## 
## Number of nodes: 376
## Number of edges: 13590
## 
## Running Louvain algorithm...
## Maximum modularity in 10 random starts: 0.7630
## Number of communities: 5
## Elapsed time: 0 seconds</code></pre>
<p>#Finally, let’s take a look at the clusters in a 2-D space</p>
<pre class="r"><code>DimPlot(object = tissue, reduction = &quot;umap&quot;)</code></pre>
<p><img src="scRNAseq_files/figure-html/seurat5-1.png" width="672" /></p>
<p>Each dot in the plot is a cell and the colour tells us to which cluster belongs. Please note that we can <em>visualize</em> cells (which are composed of 20,000 genes!) in a two-dimensional space because all the data processing we did.</p>
<p>UMAP is a powerfull clustering and visualization tool and can be used in any kind of dataset.</p>
<p>The next question is: what is the composition of the clusters? What is the identity of those cells?</p>
</div>
<div id="find-marker-genes" class="section level3">
<h3>Find marker genes</h3>
<p>Now that we foud the clusters in the data we can ask: which genes are expressed in those clusters? What makes the clusters unique and different from the others? These are important question when we want to find out more about the cell types in the sample or when looking at a sample whose composition we don’t even know e.g. a tumor sample. A marker is a gene that is specifically expressed in a cell type (or cluster). Therefore, those genes are used by biologist to classify or investigate different cells.</p>
<p>Seurat includes a function called <code>find_all_markers</code> that does exactly that.</p>
<pre class="r"><code>tissue.markers = FindAllMarkers(tissue,only.pos = T)</code></pre>
<pre><code>## Calculating cluster 0</code></pre>
<pre><code>## Calculating cluster 1</code></pre>
<pre><code>## Calculating cluster 2</code></pre>
<pre><code>## Calculating cluster 3</code></pre>
<pre><code>## Calculating cluster 4</code></pre>
<pre class="r"><code>head(tissue.markers)</code></pre>
<pre><code>##               p_val avg_logFC pct.1 pct.2    p_val_adj cluster   gene
## Emcn   5.085915e-65  3.823163 0.900 0.039 1.187103e-60       0   Emcn
## Ppap2a 1.020890e-64  3.777080 0.933 0.078 2.382860e-60       0 Ppap2a
## Flt1   7.554842e-62  2.817337 0.867 0.035 1.763376e-57       0   Flt1
## Tm4sf1 1.104579e-60  3.246169 0.892 0.070 2.578198e-56       0 Tm4sf1
## Eng    8.614232e-60  3.328788 0.917 0.109 2.010648e-55       0    Eng
## Cdh5   1.306423e-59  1.054711 0.808 0.008 3.049321e-55       0   Cdh5</code></pre>
<p>We can sort the data now using <code>dplyr</code> function <a href="https://dplyr.tidyverse.org/reference/top_n.html">top_n()</a></p>
<pre class="r"><code># argument wt in top_c() selects which variable we want to consider for choosing the top n genes
tissue.markers %&gt;% group_by(cluster) %&gt;% top_n(n = 3, wt = avg_logFC)</code></pre>
<pre><code>## # A tibble: 15 x 7
## # Groups:   cluster [5]
##       p_val avg_logFC pct.1 pct.2 p_val_adj cluster gene  
##       &lt;dbl&gt;     &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt; &lt;fct&gt;   &lt;chr&gt; 
##  1 5.09e-65      3.82 0.9   0.039  1.19e-60 0       Emcn  
##  2 1.02e-64      3.78 0.933 0.078  2.38e-60 0       Ppap2a
##  3 1.65e-55      4.04 0.792 0.027  3.84e-51 0       Plvap 
##  4 5.83e-42      2.38 0.728 0.077  1.36e-37 1       Spp2  
##  5 6.04e-39      2.63 0.621 0.033  1.41e-34 1       Slc5a2
##  6 8.30e-37      2.62 0.951 0.579  1.94e-32 1       Gpx3  
##  7 1.09e-54      3.93 0.742 0.013  2.55e-50 2       Egf   
##  8 2.27e-49      5.91 0.823 0.061  5.29e-45 2       Umod  
##  9 9.81e-38      4.62 0.694 0.061  2.29e-33 2       Klk1  
## 10 2.91e-43      3.05 0.94  0.123  6.80e-39 3       Napsa 
## 11 6.72e-31      2.93 0.78  0.126  1.57e-26 3       Kap   
## 12 8.79e-27      2.80 0.9   0.304  2.05e-22 3       Cndp2 
## 13 8.91e-29      4.42 0.854 0.185  2.08e-24 4       H2-Eb1
## 14 1.16e-27      4.49 0.902 0.23   2.71e-23 4       H2-Aa 
## 15 2.84e-26      4.38 0.927 0.301  6.63e-22 4       Cd74</code></pre>
<p>We have the genes names so we can now find what they do using a website like <a href="https://www.genecards.org/">GeneCards</a> for gene <a href="https://www.genecards.org/cgi-bin/carddisp.pl?gene=GPX3&amp;keywords=Gpx3">Gpx3</a></p>
</div>
<div id="finding-the-expression-patters-of-genes-in-a-cluster" class="section level3">
<h3>Finding the expression patters of genes in a cluster</h3>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
