<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Alejandro Granados" />


<title>A pipeline to analyse single-cell RNA sequencing data</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="index.html">CdC2019</a>
</li>
<li>
  <a href="index.html">Projects</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Workshop
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Intro.html">Day 2: Intro</a>
    </li>
    <li>
      <a href="scRNAseq.html">Day 3: scRNA seq</a>
    </li>
    <li class="divider"></li>
    <li>
      <a href="metadata.html">Additional Material</a>
    </li>
  </ul>
</li>
<li>
  <a href="https://github.com/rstudio/distill">
    <span class="fab fa fab fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">A pipeline to analyse single-cell RNA sequencing data</h1>
<h4 class="author">Alejandro Granados</h4>
<h4 class="date">24/07/2019</h4>

</div>


<div id="single-cell-rna-seq" class="section level2">
<h2>Single cell RNA seq</h2>
<p>Single cell examines the sequence information of RNA inside individual cells. By amplifying the signal from RNA inside the cell and converting the RNA into DNA, we can use sequencing methods to recover the information of which RNAs were present in a given cell. Moreover, different technologies allow to sequence multiple cells in parallel. For example, the 10x Genomics platform enables up to 10,000 cells per sample!</p>
<p>The amount of genes in a given cell (~20,000 for mouse) times the number of cells means that we are dealing with really large matrices. One challenge, however, is that until today we can not read ALL the RNA from a single cell. In the processes of sequencing, most of the RNA is lost! This <code>dropout</code> problem imposes serious challenges for data analysis.</p>
<p>Here, we will go through a standar pipeline to analyse scRNA-seq data from a cell atlas: the Tabula Muris <a href="https://tabula-muris.ds.czbiohub.org/" class="uri">https://tabula-muris.ds.czbiohub.org/</a>. In a cell atlas, researchers try to sequence the whole organisms with single cell resolution. As you can imagine, this is a huge amount of data. For the purpose of this tutorial, we will focus on a single tissue.</p>
<p>We will use the specialized library <code>Seurat</code> which has been developed by Rahul Sajita and co-workers <a href="https://satijalab.org/seurat/" class="uri">https://satijalab.org/seurat/</a></p>
<p>Let’s start by downloading the dataset from here: <a href="https://figshare.com/articles/Single-cell_RNA-seq_data_from_Smart-seq2_sequencing_of_FACS_sorted_cells_v2_/5829687" class="uri">https://figshare.com/articles/Single-cell_RNA-seq_data_from_Smart-seq2_sequencing_of_FACS_sorted_cells_v2_/5829687</a></p>
<pre class="r"><code># make sure to instal the libraries if you haven&#39;t
library(Seurat)
library(dplyr)</code></pre>
<pre><code>## 
## Attaching package: &#39;dplyr&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     filter, lag</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     intersect, setdiff, setequal, union</code></pre>
<pre class="r"><code>library(Matrix)
library(stringr)
library(readr)
library(here)</code></pre>
<pre><code>## here() starts at /Users/alejandrog/DataScienceMX</code></pre>
</div>
<div id="loading-data" class="section level2">
<h2>Loading Data</h2>
<p>We can read the data using the function <code>read.csv</code> and convert it into a sparse matrix (data is full of zeros!)</p>
<p>Sort the cells by name (their name contains information about the experimental conditions)</p>
<p>Now, let’s load the <em>metadata</em>, which contains the information about the data!</p>
<pre><code>##   plate.barcode mouse.id  tissue subtissue FACS.selection mouse.sex
## 1       D041914    3_8_M Bladder                 Multiple         M
## 2       D042253    3_9_M Bladder                 Multiple         M
## 3     MAA000487   3_10_M Bladder                 Multiple         M
## 4       B000610   3_56_F Bladder                 Multiple         F
## 5       B002764   3_38_F Bladder                 Multiple         F
## 6       B002771   3_39_F Bladder                 Multiple         F</code></pre>
<p>The <em>metadata</em> has all the information about the experiments. Every organ was processed and sequenced by a different research group and so we need to find <strong>which cells correspond to which experiment</strong>. In the <a href="https://www.nature.com/articles/s41586-018-0590-4">protocol</a> they use for scRNA-seq, each cell is sequenced on a well of a <a href="https://en.wikipedia.org/wiki/Microplate">plate</a>. The information in the metadata represents plates! Let’s link the <strong>plates</strong> to the <strong>cells</strong></p>
<p>The ERCC sequences tell us about the library size and differences between cell to cell. We can plot the amount of ERCC in the data per cell and see if there are clear outliers that we should remove.</p>
<pre class="r"><code>hist(percent.ercc)</code></pre>
<p><img src="scRNAseq_files/figure-html/ercc-1.png" width="672" /></p>
<p>Since a cell in principle should not have too many ERCC (these are only for control) we should remove cells with high ERCC content</p>
<p>ERCC sequences are useful as a <em>control</em> for the data because they were added by the scientists in <em>known</em> concentrations. In this way we expect ALL cells to have the same ERCC content, right? But things are not that simple, the technology is not perfect and different cells are sequenced in different ways or with different efficiency. Can you think of a way in which we could use ERCC sequences to <em>normalize</em> the data?</p>
<p>First, we want to get rid of cell with <em>too</em> high ERCC content. Where would you put a <em>cutoff</em>?</p>
<pre class="r"><code>raw.data = raw.data[,percent.ercc&lt;0.15]</code></pre>
<pre class="r"><code>#Take from meta.data only cells that remain in the matrix

cell.meta.filter = cell.meta.data[colnames(raw.data), ]

head(cell.meta.filter)</code></pre>
<pre><code>##                          plate.barcode mouse.id tissue subtissue
## A1.B001717.3_38_F.1.1          B001717   3_38_F Kidney          
## A10.B002775.3_39_F.1.1         B002775   3_39_F Kidney          
## A10.MAA000752.3_10_M.1.1     MAA000752   3_10_M Kidney          
## A11.MAA000801.3_11_M.1.1     MAA000801   3_11_M Kidney          
## A12.B001717.3_38_F.1.1         B001717   3_38_F Kidney          
## A12.MAA000801.3_11_M.1.1     MAA000801   3_11_M Kidney          
##                          FACS.selection mouse.sex
## A1.B001717.3_38_F.1.1            Viable         F
## A10.B002775.3_39_F.1.1           Viable         F
## A10.MAA000752.3_10_M.1.1         Viable         M
## A11.MAA000801.3_11_M.1.1         Viable         M
## A12.B001717.3_38_F.1.1           Viable         F
## A12.MAA000801.3_11_M.1.1         Viable         M</code></pre>
<p>Let’s now check how different are cells with each other. Are there cells that have not enough data? The columns in the matrix represent the cells so we can sum over the columns and plot a histogram.</p>
<pre class="r"><code># Total number of reads per cell
hist(colSums(raw.data))</code></pre>
<p><img src="scRNAseq_files/figure-html/ercc3_1-1.png" width="672" /></p>
<pre class="r"><code># But some genes might have many reads while some other genes might have low expression in living cells. 
# We can also ask then, how many genes have at least 1 read 

hist(colSums(raw.data&gt;0))</code></pre>
<p><img src="scRNAseq_files/figure-html/ercc3_1-2.png" width="672" /></p>
<p>What is the average number of genes expressed per cells?</p>
<pre class="r"><code>#Note that we define a gene as expressed if the cell has at least 1 read. But this definition can change. 

mean(colSums(raw.data&gt;0))</code></pre>
<pre><code>## [1] 1665.79</code></pre>
<pre class="r"><code>#Based on the histogram we can keep only cells that have more than 1000 genes with &gt;0 reads
raw.data&lt;-raw.data[,colSums(raw.data&gt;0)&gt;1000]
#Try looking at the histogram now</code></pre>
<p>Anothe imporant filter is by gene. Some genes are very poorly expressed, maybe because they have low expression in general or maybe they are not expressed in this particular tissue/condition.</p>
<p>To see the distribution of genes, we need to sum by row.</p>
<pre class="r"><code># Number of reads for each gene across cells
hist(rowSums(raw.data))</code></pre>
<p><img src="scRNAseq_files/figure-html/filtergenes-1.png" width="672" /></p>
<pre class="r"><code># We can also look at how many cells express each gene
hist(rowSums(raw.data&gt;0))</code></pre>
<p><img src="scRNAseq_files/figure-html/filtergenes-2.png" width="672" /></p>
<div id="seurat" class="section level3">
<h3>Seurat</h3>
<p>Seurat is a R-based software package for analysis of single cell RNA seq. I has been shown to perfom well across different datasets and normalization methods.</p>
<p>Here are some references where people has compared Seurat with other methods (there are many methods and packages for analysis of single cell RNA seq!) <a href="https://www.nature.com/articles/s41592-019-0425-8" class="uri">https://www.nature.com/articles/s41592-019-0425-8</a></p>
<pre class="r"><code># # # # ## SEURAT OBJECT
tissue &lt;- CreateSeuratObject(counts = raw.data)</code></pre>
<pre><code>## Warning: Feature names cannot have underscores (&#39;_&#39;), replacing with dashes
## (&#39;-&#39;)</code></pre>
<pre class="r"><code>tissue &lt;- AddMetaData(object = tissue, cell.meta.data)
tissue &lt;- AddMetaData(object = tissue, percent.ercc, col.name = &quot;percent.ercc&quot;)</code></pre>
<p>We can now normalize the data. One of the technical problems of RNA sequencing is that some RNA molecules are not detected and other are highly produced in cells, so the <em>distribution</em> of expression is <em>wide</em>.</p>
<pre class="r"><code>  tissue&lt;-NormalizeData(object = tissue, verbose=F)
  tissue&lt;-FindVariableFeatures(object=tissue,selection.method =&quot;vst&quot;,nfeatures =2000)
   tissue&lt;-ScaleData(tissue)</code></pre>
<pre><code>## Centering and scaling data matrix</code></pre>
<pre class="r"><code>tissue &lt;- RunPCA(object = tissue, npcs = 100, verbose = FALSE)

VizDimLoadings(object = tissue, dims = 1:4, reduction = &quot;pca&quot;)</code></pre>
<p><img src="scRNAseq_files/figure-html/seurat3-1.png" width="672" /></p>
<pre class="r"><code>ElbowPlot(tissue)</code></pre>
<p><img src="scRNAseq_files/figure-html/seurat3-2.png" width="672" /></p>
<pre class="r"><code>#Run this only if you have python installed and if you installed UMAP
tissue &lt;- RunUMAP(object = tissue, reduction = &quot;pca&quot;, dims = 1:40)

# A different method for visualization is tSNE
tissue &lt;- RunTSNE(tissue,dims=1:10)

tissue &lt;- FindNeighbors(object = tissue, reduction = &quot;pca&quot;, dims = 1:40)</code></pre>
<pre><code>## Computing nearest neighbor graph</code></pre>
<pre><code>## Computing SNN</code></pre>
<pre class="r"><code>tissue &lt;- FindClusters(tissue, resolution = 0.8)</code></pre>
<pre><code>## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
## 
## Number of nodes: 376
## Number of edges: 13590
## 
## Running Louvain algorithm...
## Maximum modularity in 10 random starts: 0.7630
## Number of communities: 5
## Elapsed time: 0 seconds</code></pre>
<p>#Finally, let’s take a look at the clusters in a 2-D space</p>
<pre class="r"><code>DimPlot(object = tissue, reduction = &quot;umap&quot;)</code></pre>
<p><img src="scRNAseq_files/figure-html/seurat5-1.png" width="672" /></p>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
