---
title: "Making an interactive 3D UMAP plot"

knit: (function(input_file, encoding) {
  out_dir <- '.';
  rmarkdown::render(input_file,
 encoding=encoding,
 output_file=file.path(dirname(input_file), out_dir, 'UMAP_3d.html'))})

author: "Alejandro A Granados"
date: "13/04/2020"
output: html_document

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


 
 
## An interactive 3D UMAP plot using plotly.

<br>
UMAP plots have become the standard for representing single-cell RNA seq data. If you are not using UMAP and still using tSNE, you should stop. Right now. UMAP plots succesfully represent not only the clusters in the high-dimensional space but also the *relationships* between the clusters. Which is really important when analysing development datasets. 
<br>


```{r loadlib, echo=T, message=F, warning=F}
library(plotly)
library(dplyr)
library(knitr)
```


```{r hidden_process, echo=F, message=F, warning=F}

load("/Users/alejandrog/Downloads/UMAP_coordinates_annotated.rda")

plot.data.ann %>% filter(barcode %in% sample(plot.data.ann$barcode,1000))->scatter.data
scatter.data %>% select(barcode,CellType, age, UMAP_1,UMAP_2,UMAP_3, seurat_clusters) ->scatter.data
```
Our data: a single-cell RNA seq [dataset](https://www.sciencedirect.com/science/article/pii/S0896627319303459) where observations are individual cells and the variables are cell's features such as cell type (annotated by authors), the embryo age from which the cell comes from and the UMAP coordinates (which I generated for this example using a custom RNA-seq pipeline). Data also includes the variable seruat_clusters which was generated by clustering cells at the transcriptome level using `Seurat`.

```{r table, echo=T, message=F, warning=F}
kable( head(scatter.data))
```

The question we would like to answer with the interactive plot is if the UMAP captures the biological relationships of retinal development. To achieve that we could use an interactive UMAP plot annotated by cell type and age. 

But first things first: colors! `Set3` is pretty decent qualitative palette that I prefer over the other ones in `RColorBrewer` because it's pretty and to my eyes has the most distinct set of colors among qualitative palettes. 

```{r makeColors, echo=T, message=F, warning=F}

newCols<-colorRampPalette(RColorBrewer::brewer.pal(12,'Set3')) 
# We want a different color for each cell type
mycolors = newCols(scatter.data$CellType %>% unique() %>% length() )

```

And wihtout further ado, our interactive UMAP using plotly: 

```{r makePlot, echo=T, message=F, warning=F}


fig <- plot_ly(scatter.data, x = ~UMAP_1, y = ~UMAP_2, z = ~UMAP_3,                
    marker = list(size = 3), color = ~CellType,                 
    colors = mycolors,                
    text=~paste("CellType:",CellType,"Age:",age,"Cluster:",seurat_clusters),                
    hoverinfo = 'text') 

fig %>% add_markers()
```

As you can see plotly takes the `color` parameter which can be mapped to a variable in the dataframe. We also provide `colors` for each cell type. The `text` argument it's very interesting because with can include variable names in a `paste` which creates a complex label for each cell. 